delegation/src/Pos/Arbitrary/Delegation.hs

```diff
@@ -7,27 +7,28 @@ module Pos.Arbitrary.Delegation
 import           Universum
 
 import qualified Data.HashMap.Strict as HM
-import           Test.QuickCheck (Arbitrary (..), Gen, listOf)
+import           Serokell.Util (allDistinct)
+import           Test.QuickCheck (Arbitrary (..), Gen, listOf, suchThat)
 import           Test.QuickCheck.Arbitrary.Generic (genericArbitrary, genericShrink)
 
 import           Pos.Arbitrary.Core ()
 import           Pos.Binary.Core ()
-import           Pos.Core (EpochIndex, HasConfiguration)
+import           Pos.Core (EpochIndex, HasConfiguration, HeavyDlgIndex (..))
 import           Pos.Crypto (ProxySecretKey (..), createPsk)
-import           Pos.Delegation.Types (DlgPayload (..), DlgUndo)
+import           Pos.Delegation.Types (DlgPayload (..), DlgUndo (..))
 
 genDlgPayload :: HasConfiguration => EpochIndex -> Gen DlgPayload
 genDlgPayload epoch =
-    UnsafeDlgPayload . toList . HM.fromList . map convert <$>
-    listOf genPSK
+    UnsafeDlgPayload . toList . HM.fromList . map convert <$> listOf genPSK
   where
     convert psk = (pskIssuerPk psk, psk)
-    genPSK = createPsk <$> arbitrary <*> arbitrary <*> pure epoch
+    genPSK = createPsk <$> arbitrary <*> arbitrary <*> pure (HeavyDlgIndex epoch)
 
 instance HasConfiguration => Arbitrary DlgPayload where
     arbitrary = arbitrary >>= genDlgPayload
     shrink = genericShrink
 
 instance HasConfiguration => Arbitrary DlgUndo where
-    arbitrary = genericArbitrary
+    arbitrary =
+        genericArbitrary `suchThat` (allDistinct . map pskIssuerPk . duPsks)
```

Comment:
Ideally it should be written without `suchThat`, because it's likely inefficient. You can generate a list of PSKs with different issuers quite easily (e. g. create a list and use some kind of `nub` which takes a function).

Reply:
I think this change is not important anyway, I've reverted it.